### DISCUSSION TITLE
Would a Pipenv / pytest contribution be welcome?

### DISCUSSION THREAD

**Discussion by TechnicalPirate** [marcelotduarte/cx_Freeze#1186] [Q&A] [answered]
Hi! 

Quick question, as I'm looking at contributing to help resolve this issue : https://github.com/marcelotduarte/cx_Freeze/issues/594

I just started looking at the repo structure, it seems most the libraries used are "out of the box" python modules. 

Would there be interest in having a PR contributing a basic [pipenv](https://pipenv.pypa.io/en/latest/) setup so we could use the [pytest ](https://docs.pytest.org/en/6.2.x/)framework or is the expectation to only work with the standard library?

Cheers

:)

**Comment by marcelotduarte** [comment]
Hi!
I think the author only wanted to use the standard library. I try to follow that line if possible. But I use importlib-metadata which is not standard on py36 and py37 for example. Even on py38 and py39 the version I use has more features than the default. So yes, if it is beneficial, I accept to use.
I have a CI that tests some of the [samples](https://github.com/marcelotduarte/cx_Freeze/tree/main/cx_Freeze/samples). I'm still building these tests, because I'm trying to test if I can build and run the sample, and I don't use pytest.
In some cases, I test without venv and with venv (python -m venv) and I have a test with pipenv for some cases that I knew would give an error to correct them.
In conclusion, the use of pytest is welcome. Alias, the original author or some contributor started using some type of [test](https://github.com/marcelotduarte/cx_Freeze/tree/main/test)
There is an issue about using pytest #747

**Reply by TechnicalPirate** [reply]
Ok cool, 

I understand the aversion to none-standard modules in the _product_ itself `cx_freeze`- that's a reasonable approach.

_

I think usage of `pipenv` would be beneficial while not contradicting the spirit of the above as it supports dedicated `[dev-requirements]` which would allow usage of other packages ( like `pytest` ) to aid in development without polluting the final product :) 

Regarding `pytest` itself - After working with the standard `unittest` module and `pytest` i can work with both but at this point have far more familiarity with `pytest` ( and love the fact you can spit out a `html` coverage report! ) 

With your blessing my approach to the linked issue would be:

* Fork + add `pipfile` + `pyenv` to dev-deps
* Reach test-coverage parity with what already exists
* _At this point optionally submit a PR with the above ^_
* Implement a set of tests to enforce Apples expected app structure
* Rework `bdist_mac` logic to utilize the info provided in the gists on the linked issue
* _Submit PR / Review_

Let me know your thoughts and i'll look to get the first bits done in the coming weeks 

Cheers! 😸

**Reply by marcelotduarte** [reply]
It took me a while to answer because I talked to @anthony-tuininga  that says:
I think that is an ideal situation but by no means necessary. If something is available that does exactly what you want it to do, why not use it? On the other hand, it should provide a significant benefit -- not just eliminate a couple of lines of code!

I would also like you to read if you haven't already: https://cx-freeze.readthedocs.io/en/latest/development/index.html#contributing

That is, after "Reach test-coverage parity with what already exists" you must submit a PR.

You should also consider that we use cx_Freeze on several operating systems: Linux, macOS, Windows, as well as other Unix-like ones like FreeBSD.
It is also used in various virtual environments (venv, virtualenv, pipenv) and various distributions, such as MSYS2, conda, macPorts...

**Reply by TechnicalPirate** [reply]
Hi @marcelotduarte 

So finally getting to move forward with some of this :) 

I forked and pulled the repo today and opened it as a new PyCharm project - my project interpreter configured to a `pipenv` environment created from a `Python3.9` interpreter

I'll be logging my progress / issues encountered here to keep a log / hopefully make sense of changes I'll make.

So far - with the goal to just "run whats there" - I'm getting failures within the `test_finder`:

* `importlib_metadata` is not found ( appears to have been renamed to `importlib.metadata` as of Python 3.8? ) 🛠️ - Just mass renamed imports locally as a work around
* `nose` is not found - 🛠️  to manually `pip` install to work around for now
*  Cyclic import?
```
from . import util as winutil
ImportError: cannot import name 'util' from partially initialized module 'cx_Freeze' (most likely due to a circular import) (C:\King\repos\git\cx_Freeze\cx_Freeze\__init__.py)
```
 seems like the module just isn't available and needs compiling *into* the package? 

... is the expectation to build a `.whl` then execute the tests against that? :)

**Reply by marcelotduarte** [reply]
>     * `importlib_metadata` is not found ( appears to have been renamed to `importlib.metadata` as of Python 3.8? ) hammer_and_wrench - Just mass renamed imports locally as a workaround

importlib_metadata works with all python versions, importlib.metadata works only on py>=3.8 but has bugs, so I use importlib_metadata in all python versions. It can be installed with pip.
 
>     * Cyclic import?

cx_Freeze.util is a extension module compiled when you install cx_Freeze in windows. In others systems it is not needed.

after a git, you can install the cx_Freeze in your system with: 
pip install -e .
which installs the cx_Freeze and the dependencies like importlib_metadata

**Reply by TechnicalPirate** [reply]
Cool - I'll update my local env to install `importlib_metadata` then as a package dependency. 

Re: the `util` module - how do you handle generating this for the unit-test suite? / would be good if you can share your current test-running configuration on windows as "out of the box" ( just cloning the repo and running tests ) will fail as this isn't generated.

**Reply by TechnicalPirate** [reply]
Ok bit more progress today :) 

1) On my fork i've added basic `pipenv` setup with these dependencies
```
[packages]
importlib_metadata = ">=3.1.1"

[dev-packages]
nose = "*"
pygments = "*"
```
2) Realized to avoid the `utils` import error on test execution i must run `pip install -e .` prior to the test suite
3) Now when running the test suite i get no import errors... *but* i also get no tests discovered / executed? 

Any ideas? :) 

edit: - I'm blind - needed to use `nose` as the runner - now its discovering the tests :)

**Reply by TechnicalPirate** [reply]
Ok @marcelotduarte  - i think i have things setup correctly now but am seeing a number of test failures....

![image](https://user-images.githubusercontent.com/35609336/134930226-1af97b56-30bc-4aa4-94b3-f98c980821be.png)
I'll look to investigate + fix these locally but it would be good to confirm if these are passing/failing on other windows machines

In the meantime - as requested in the contributors guide - many small PR's are preferable - [so here's my first :)](https://github.com/marcelotduarte/cx_Freeze/pull/1224)

**Reply by marcelotduarte** [reply]
I didn't pay attention to the idea that you only use pipenv. When I started contributing 2 or 3 years ago, I didn't have any problems, as I use direct pip.
The file setup.cfg section options have all the requirements, so if you run `pip install -e .` installs the requirements and package initially. And if you have changed the code in C, you have to uninstall and install again, to generate the extensions again. But don't be the case with your proposal.

The relevant part in setup.cfg
```
[options]
python_requires = >=3.6
install_requires =
    cx-logging >= 3.0 ; sys_platform == "win32"
    importlib-metadata >= 4.3.1
zip_safe = False
```

**Reply by marcelotduarte** [reply]
Since I took over maintenance, I've practically not used the tests that were already there. I test in a different way, as I said [previously](https://github.com/marcelotduarte/cx_Freeze/discussions/1186#discussioncomment-1094882). So, as I've been changing the API, there must be errors to correct in the tests.

**Reply by TechnicalPirate** [reply]
Ok no worries my goal was to get things to a "stable" place before i started switching testing frameworks.

if the tests are failing due to pre-existing issues i don't need to worry as much then and i can move onto the next step of switching to the pytest framework.

edit: Re: `pipenv` - i don't exclusively use this - i can use `pip` if required, but i prefer `pipenv` due to it being deterministic AND handling `venv` construction / teardown. 

edit++: good to know the `setup.cfg` handles getting `importlib_metadata` i'll remove that from the `pipfile` so it just contains developer-dependencies

**Reply by marcelotduarte** [reply]
In PR #1226 I added a stub for util module, to satisfy IDEs.
In PR #1230 I explicitly put `pip install -e .` in the development section.

**Reply by TechnicalPirate** [reply]
Hey mate :) 

Thanks for the updates - I had some more time today to poke around more.

I've fixed up the existing tests bar one - [PR Here](https://github.com/marcelotduarte/cx_Freeze/pull/1234) for your review.

I'll be on vacation next week so *less* able to contribute - In the interim please let me know if you want any changes making to the above PR or my prior one for the `gitIgnore` / `pipenv` [here](https://github.com/marcelotduarte/cx_Freeze/pull/1224)

Cheers

:)

**Reply by TechnicalPirate** [reply]
@marcelotduarte - Hey mate, back from vacations + caught up on production work, available to continue my contributions

_( Fix existing tests > switch to pytest > write tests for OSX package structure > refactor OSX packaging code )_

Can you let me know how to proceed? I submitted 2 PR's already to get everything in place to start switching to Pytest but i need them reviewing / approving before i can continue ( without creating mammoth PR's at least! ) 

* Change dep-management :  https://github.com/marcelotduarte/cx_Freeze/pull/1224
* Fix existing tests : https://github.com/marcelotduarte/cx_Freeze/pull/1234


_Edit: Wanted to move forward with this so prepared a *draft* pull request where i've already done the conversion to the new framework - you can get it / try it [here](https://github.com/marcelotduarte/cx_Freeze/pull/1255)_

Current coverage is 27% 
![image](https://user-images.githubusercontent.com/35609336/134929747-e5f2998f-6529-4fba-be59-46d6494cb65a.png)


Cheers

:)

**Reply by marcelotduarte** [reply]
Hi! I'm short on time these days, but I'll try to make time so I don't get in the way of your work progress. I'll answer you in PRs.

**Reply by TechnicalPirate** [reply]
Heya :) 

No worries, very understanding on my end :) 

Once we've discussed / moved forward with the current Open PR's my focus will be "write tests for OSX package structure" + the _real_ contribution after that! .... i think we're offset 12 hours from each other IRL anyway so its a bit of a tag-team affair regardless :)

**Comment by marcelotduarte** [comment]
> 'll look to investigate + fix these locally but it would be good to confirm if these are passing/failing on other windows machines

Are you using windows?

**Reply by TechnicalPirate** [reply]
Yes :) but i can also use a Mac - Depending on how far i / we go with this i was going to setup a job on our internal CI as well to run cross-platform tests.

**Comment by TechnicalPirate** [comment, marked as answer]
After discussion with maintainer on some PR's the outcome of this was 

1) We should *NOT* change package/venv manager - the project will stick with `pip` and `venv` 
2) We *should* generate a `dev-requirements.txt`
3) We *can* switch testing framework from `nose` to `pytest`

I will mark this outcome as the answer to this question - further discussion on the nose > pytest switchover will happen on this PR : https://github.com/marcelotduarte/cx_Freeze/pull/1255
